backend/src/test/java/ai/verta/modeldb/ProjectTest.java

```diff
@@ -28,41 +38,50 @@
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Random;
 import java.util.Set;
 import java.util.UUID;
+import java.util.concurrent.ExecutionException;
 import java.util.stream.Collectors;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.junit.After;
 import org.junit.Before;
-import org.junit.FixMethodOrder;
 import org.junit.Test;
 import org.junit.runner.RunWith;
-import org.junit.runners.JUnit4;
-import org.junit.runners.MethodSorters;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringRunner;
 
-@RunWith(JUnit4.class)
-@FixMethodOrder(MethodSorters.NAME_ASCENDING)
-public class ProjectTest extends TestsInit {
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = App.class, webEnvironment = DEFINED_PORT)
+@ContextConfiguration(classes = {ModeldbTestConfigurationBeans.class})
+public class ProjectTest extends ModeldbTestSetup {
```

Comment:
it is not just a hierarchy we have `@Autowired` annotations there for commonly used 
```
  @Autowired UAC uac;
  @Autowired TestConfig testConfig;
  @Autowired Executor executor;
  @Autowired ReconcilerInitializer reconcilerInitializer;
``` 
https://github.com/VertaAI/modeldb/pull/3228/files#diff-4cb455ea3bf6f37aed6c6cbb3abc603bf548cb87cdcc4165defc4de826d71e78R52

We have used `@SpringBootTest` so we have used the `@Autowired` functionality for mocking UAC and other beans.

Reply:
We have defined many commonly used codes in ModeldbTestSetup which we have to initialize on each test for ex: beans, mock UAC bean, all *Stub (entire initializedChannelBuilderAndExternalServiceStubs, setupMockUacEndpoints, cleanUpResources methods), etc.
```
  @Autowired UAC uac;
  @Autowired TestConfig testConfig;
  @Autowired Executor executor;
  @Autowired ReconcilerInitializer reconcilerInitializer;
```  
If we are not used hierarchy then we have to define all the above things in each test class and it will duplicate the code.

Reply:
I would argue that a little bit of duplicated code is far better than not being able to easily read code and understand where methods are coming from in a class hierarchy. This is ok for now, but we should absolutely _stop_ creating class hierarchies just for code reuse. This is not a good reason for using inheritance, as there are other, less confusing ways to reuse code than to use a class hierarchy.